modules/index/src/main/java/org/apache/ignite/internal/index/ChangeIndexStatusTaskScheduler.java

```diff
@@ -0,0 +1,214 @@
+        if (!closeGuard.compareAndSet(false, true)) {
+            return;
+        }
+
+        busyLock.block();
+
+        stopAllTasks();
+    }
+
+    /**
+     * Schedules a task for that will transfer the given index to the {@link CatalogIndexStatus#BUILDING} state.
+     */
+    void scheduleStartBuildingTask(CatalogIndexDescriptor indexDescriptor) {
+        assert indexDescriptor.status() == CatalogIndexStatus.REGISTERED;
+
+        LOG.info("Scheduling starting of index building. Index: {}", indexDescriptor);
+
+        inBusyLockSafe(busyLock, () -> {
+            var taskId = new ChangeIndexStatusTaskId(indexDescriptor);
+
+            scheduleTask(taskId, startBuildingIndexTask(indexDescriptor));
+        });
+    }
+
+    private ChangeIndexStatusTask startBuildingIndexTask(CatalogIndexDescriptor indexDescriptor) {
+        return new ChangeIndexStatusTask(
+                indexDescriptor,
+                catalogManager,
+                placementDriver,
+                clusterService,
+                logicalTopologyService,
+                clock,
+                clockWaiter,
+                executor,
+                busyLock
+        ) {
+            @Override
+            CatalogCommand switchIndexStatusCommand() {
+                return StartBuildingIndexCommand.builder().indexId(indexDescriptor.id()).build();
+            }
+        };
+    }
+
+    /**
+     * Schedules a task for that will remove a given index from the Catalog.
+     */
+    void scheduleRemoveIndexTask(CatalogIndexDescriptor indexDescriptor) {
+        assert indexDescriptor.status() == CatalogIndexStatus.STOPPING;
+
+        LOG.info("Scheduling index removal. Index: {}", indexDescriptor);
+
+        inBusyLockSafe(busyLock, () -> {
+            var taskId = new ChangeIndexStatusTaskId(indexDescriptor);
+
+            scheduleTask(taskId, removeIndexTask(indexDescriptor));
+        });
+    }
+
+    private ChangeIndexStatusTask removeIndexTask(CatalogIndexDescriptor indexDescriptor) {
+        return new ChangeIndexStatusTask(
+                indexDescriptor,
+                catalogManager,
+                placementDriver,
+                clusterService,
+                logicalTopologyService,
+                clock,
+                clockWaiter,
+                executor,
+                busyLock
+        ) {
+            @Override
+            CatalogCommand switchIndexStatusCommand() {
+                return RemoveIndexCommand.builder().indexId(indexDescriptor.id()).build();
+            }
+        };
+    }
+
+    private void scheduleTask(ChangeIndexStatusTaskId taskId, ChangeIndexStatusTask task) {
+        // Check if the task has already been added before.
+        if (taskById.putIfAbsent(taskId, task) == null) {
+            task.start().whenComplete((unused, throwable) -> taskById.remove(taskId));
+        } else {
+            LOG.info("Skipping task scheduling, because a task with the same ID is already running.");
```

Comment:
I would take it to the debug level.

Reply:
This message should not appear often while it may help to locate a race

Reply:
ok